lib/elastic_apm/transport/base.rb

```diff
@@ -8,90 +8,156 @@
 
 module ElasticAPM
   module Transport
+    # rubocop:disable Metrics/ClassLength
     # @api private
     class Base
       include Logging
 
+      WATCHER_EXECUTION_INTERVAL = 5
+      WATCHER_TIMEOUT_INTERVAL = 4
+      WORKER_JOIN_TIMEOUT = 5
+
       def initialize(config)
         @config = config
         @queue = SizedQueue.new(config.api_buffer_size)
-        @pool = Concurrent::FixedThreadPool.new(config.pool_size)
-        @workers = []
 
         @serializers = Serializers.new(config)
         @filters = Filters.new(config)
+
+        @stopped = Concurrent::AtomicBoolean.new
+        @workers = Array.new(config.pool_size)
+
+        @watcher_mutex = Mutex.new
+        @worker_mutex = Mutex.new
```

Comment:
Now you also need to change [mutex](https://github.com/elastic/apm-agent-ruby/blob/39488e5f7e9d3c1d3f8fa86ccf1e3661830b3b28/lib/elastic_apm/transport/base.rb#L133) to a `worker_mutex`.